notebooks/message.ipynb

```diff
@@ -0,0 +1,410 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "id": "57713f32",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "assert result.HasField(\"solution\")  # Above KnapSack problem has a solution\n",
+    "solution = result.solution"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6b92c135",
+   "metadata": {},
+   "source": [
+    "The main part of the solution is the values of the decision variables $x_i$. They are stored in `state` field:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "id": "1750459c",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "{3: 1.0, 0: 1.0, 2: 0.0, 1: 0.0, 5: 0.0, 4: 0.0}"
+      ]
+     },
+     "execution_count": 9,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "solution.state.entries"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "297a6a70",
+   "metadata": {},
+   "source": [
+    "This says $x[0] = x[3] = 1$ and others are $0$. Then the profit is $p[0] + p[3] = 10 + 31 = 41$."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "id": "0754f82f",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "41.0"
+      ]
+     },
+     "execution_count": 10,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "solution.objective"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "601b5629",
+   "metadata": {},
+   "source": [
+    "OMMX normalizes inequality constraints into $f(x) \\leq 0$ form, i.e. the above constraint is normalized as follows:\n",
+    "\n",
+    "$$\n",
+    "f(x) = \\sum_{i=1}^n w_i x_i - W \\leq 0\n",
+    "$$\n",
+    "\n",
+    "Substituting $x$, its yields $f(x) = w[0] + w[3] - W = 11 + 35 - 47 = -1$. This is stored in `evaluated_value` field."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "id": "1241cd61",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "-1.0"
+      ]
+     },
+     "execution_count": 11,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "assert len(solution.evaluated_constraints) == 1  # There is only one constraint\n",
+    "constraint = solution.evaluated_constraints[0]\n",
+    "constraint.evaluated_value"
+   ]
+  }
+ ],
+ "metadata": {
+  "kernelspec": {
+   "display_name": "Python 3 (ipykernel)",
+   "language": "python",
+   "name": "python3"
+  },
+  "language_info": {
+   "codemirror_mode": {
+    "name": "ipython",
+    "version": 3
+   },
+   "file_extension": ".py",
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.11.4"
+  }
```

Comment:
# Original

``````markdown
## Solve using Python-MIP
OMMX project manages an adapter package `ommx-python-mip-adapter`, which provides a bridge between OMMX and Python-MIP. Python-MIP is a Python library for mathematical programming, which provides a high-level modeling API. The adapter package converts `Instance` object to Python-MIP model, and vice versa.
Let's solve the above Knapsack instance using Python-MIP:

```python
import ommx_python_mip_adapter as adapter
result = adapter.solve(instance)
```

`result` is object of type `ommx.v1.Result` which contains `solution` field if solved. This may contain `infeasible` field if the problem is infeasible, i.e. the solver prove that all solution of the problem is infeasible, or `unbounded` field if the problem is unbounded.

```python
assert result.HasField("solution")  # Above KnapSack problem has a solution
solution = result.solution
```

The main part of the solution is the values of the decision variables $x_i$. They are stored in `state` field:

```python
solution.state.entries
```

This says $x[0] = x[3] = 1$ and others are $0$. Then the profit is $p[0] + p[3] = 10 + 31 = 41$.

```python
solution.objective
```

OMMX normalizes inequality constraints into $f(x) \leq 0$ form, i.e. the above constraint is normalized as follows:
$$
f(x) = \sum_{i=1}^n w_i x_i - W \leq 0
$$
Substituting $x$, its yields $f(x) = w[0] + w[3] - W = 11 + 35 - 47 = -1$. This is stored in `evaluated_value` field.

```python
assert len(solution.evaluated_constraints) == 1  # There is only one constraint
constraint = solution.evaluated_constraints[0]
constraint.evaluated_value
```
``````

# Suggestion

``````markdown
## Solving 

OMMX itself doesn't include solvers. Instead, it provides a standard format for representing problems that can be consumed by various solvers via adapters.

Let's see how to solve our Knapsack problem using the Python-MIP solver via the `ommx-python-mip-adapter` package:

```python
import ommx_python_mip_adapter as adapter

result = adapter.solve(instance)
```

The returned `result` is an `ommx.v1.Result` object containing the solution if the problem was solved successfully:

```python
assert result.HasField("solution")  
solution = result.solution
```

The main part of the solution are the values of the decision variables $x_i$, stored in the `state` field:

```python
print(solution.state.entries)
```

Here $x[0] = x[3] = 1$ and the rest are $0$, giving a total profit of $p[0] + p[3] = 10 + 31 = 41$:

```python
print(solution.objective)
```

OMMX normalizes constraints to the form $f(x) \leq 0$. For our knapsack constraint:

$$
f(x) = \sum_{i=1}^n w_i x_i - W \leq 0
$$

Plugging in the solution $x$ gives $f(x) = w[0] + w[3] - W = 11 + 35 - 47 = -1$, stored in the `evaluated_value` field:

```python
assert len(solution.evaluated_constraints) == 1  
print(solution.evaluated_constraints[0].evaluated_value)
```
``````

# My Suggestions

- Clarify that OMMX itself doesn't include solvers but provides a standard format for representing problems that can be consumed by solvers via adapters.
- Show how to solve the Knapsack problem using the Python-MIP solver via the `ommx-python-mip-adapter` package.
- Explain that the `result` is an `ommx.v1.Result` object containing the solution if the problem was solved successfully.
- Demonstrate accessing the values of the decision variables in the `state` field of the solution.
- Show how to calculate the total profit using the optimal decision variable values.
- Explain that OMMX normalizes constraints to the form $f(x) \leq 0$ and show how to access the evaluated constraint value in the solution.